### PR TITLE
Made a StructuredBuffer class to be shared by all StructuredBuffer-type resources.

### PR DESCRIPTION
**Pull request by jaremieromer**
See title.
Cutover `MaterialData` to use this type instead of having a separate `MaterialDataBufferResource` class. 

Still need to cutover `MeshRendererBufferResource` but there's some work to do there.

### CODE REVIEW

File: test/integration_test/ncengine/graphics2/StructuredBuffer_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+class StructuredBufferTest : public DiligentEngineFixture
+{
+    protected:
+        static constexpr auto signatureName = "testSignature";
+        static constexpr auto variableName = "testBuffer";
+        static constexpr auto bindingIndex = uint8_t{1};
+        static constexpr auto maxInstanceCount = 10u;
+        static constexpr auto initialInstanceCount = 3u;
+
+        std::unique_ptr<nc::graphics::MaterialResourceSignature> signature;
```
Comment (McCallisterRomer) [inline test/integration_test/ncengine/graphics2/StructuredBuffer_tests.cpp:18]: I suppose this test doesn't have to depend on `MaterialResourceSignature` or `MaterialData` anymore.

File: source/ncengine/graphics2/diligent/resource/base/StructuredBuffer.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+#include <span>
+
+namespace nc::graphics
+{
+class StructuredBufferBase
+{
+    public:
+        explicit StructuredBufferBase(std::string_view name,
+                                      Diligent::IShaderResourceVariable& variable,
+                                      uint32_t maxSize,
```
Comment (jaremieromer) [inline source/ncengine/graphics2/diligent/resource/base/StructuredBuffer.h]: change to `maxElementCount` and `initialElementCount` to be consistent with below

File: source/ncengine/graphics2/diligent/resource/base/StructuredBuffer.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                                      uint32_t initialSize);
+
+        static auto MakeResourceDesc(std::string_view variableName, Diligent::SHADER_TYPE shaderType) -> Diligent::PipelineResourceDesc;
+        void Transition(Diligent::IDeviceContext& context, Diligent::RESOURCE_STATE state);
+
+        auto GetShaderVariable()       -> Diligent::IShaderResourceVariable& { return *m_variable; }
+        auto Size()              const -> uint32_t                           { return m_size;      }
```
Comment (jaremieromer) [inline source/ncengine/graphics2/diligent/resource/base/StructuredBuffer.h]: Change to `..Count` for consistency

File: source/ncengine/graphics2/diligent/resource/base/StructuredBuffer.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+class StructuredBuffer : public StructuredBufferBase
+{
+    public:
+        static constexpr auto ElementStride = static_cast<uint32_t>(sizeof(T));
+
+        explicit StructuredBuffer(Diligent::IDeviceContext& context,
+                                  Diligent::IRenderDevice& device,
+                                  std::string_view name,
+                                  Diligent::IShaderResourceVariable& variable,
+                                  uint32_t maxElements,
```
Comment (jaremieromer) [inline source/ncengine/graphics2/diligent/resource/base/StructuredBuffer.h]: Change `maxElements` to `maxElementCount` to match `initialElementCount`

File: source/ncengine/graphics2/diligent/resource/component/ComponentResourceSignature.cpp [hunk 5]
```diff
@@ -40,11 +42,13 @@ ComponentResourceSignature::ComponentResourceSignature(Diligent::IDeviceContext&
         throw NcError{fmt::format("Failed retrieving shader variable '{}'", meshRendererBufferVariableName)};
     }
 
-    m_meshRendererResource = std::make_unique<MeshRendererBufferResource>
+    m_meshRendererResource = std::make_unique<StructuredBuffer<MeshRendererData>>
     (
         context,
         device,
+        "MeshRendererStructuredBuffer",
```
Comment (jaremieromer) [inline source/ncengine/graphics2/diligent/resource/component/ComponentResourceSignature.cpp]: If it works, lets make this `meshRendererBufferVariableName` . It makes sense to me that the name you pass into `ShaderBindings.h` (variable name) should also be what appears in the GL/HLSL

**Review by McCallisterRomer — APPROVED** [review]

**Pull request merged**